smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/waiters/JMESPathVisitor.kt

```diff
@@ -0,0 +1,232 @@
+        suffixSequence.map { "$preferredName$it" }.first(tempVars::add)
+
+    // Some JMESPath expressions may have their own valid JMESPath expressions
+    // within them, i.e. to map or filter.  This method is called to render
+    // those expressions.
+    private fun childBlock(forExpression: JmespathExpression): String =
+        forExpression.accept(JMESPathVisitor(writer))
+
+    // Maps the expression result in leftName into a new collection using the right expression.
+    private fun mappingBlock(right: JmespathExpression, leftName: String): String {
+        if (right is CurrentExpression) return leftName // Nothing to map
+        val outerName = bestTempVarName("projection")
+        writer.openBlock("let \$L = Optional.some((\$L ?? []).compactMap { current in", "})", outerName, leftName) {
+            writer.write("let current = Optional.some(current)")
+            val innerResult = addTempVar("innerResult", childBlock(right))
+            writer.write("return \$L", innerResult)
+        }
+        return outerName
+    }
+
+    // Accesses a field on the expression named parentName & returns the name of the
+    // resulting expression.
+    private fun subfield(expression: FieldExpression, parentName: String): String {
+        val name = expression.name.toLowerCamelCase()
+        return addTempVar(name, "\$L?.\$L", parentName, name)
+    }
+
+    // Performs a Boolean "and" of the left & right expressions
+    // A Swift compile error will result if both left & right aren't Bool.
+    override fun visitAnd(expression: AndExpression): String {
+        val leftExp = expression.left!!.accept(this)
+        val rightExp = expression.right!!.accept(this)
+        return addTempVar("andResult", "\$L && \$L", leftExp, rightExp)
+    }
+
+    // Perform a comparison of two values.
+    // The JMESValue type is used to provide conversion and comparison as needed between types
+    // that aren't comparable in "pure Swift" (i.e. Int to Double or String to RawRepresentable
+    // by String.)
+    // The Smithy comparator is a string that just happens to match up with all Swift comparators,
+    // so it is rendered into Swift as-is.
+    override fun visitComparator(expression: ComparatorExpression): String {
+        val left = expression.left!!.accept(this)
+        val right = expression.right!!.accept(this)
+        return addTempVar("comparison", "JMESValue(\$L) \$L JMESValue(\$L)", left, expression.comparator, right)
+    }
+
+    override fun visitCurrentNode(expression: CurrentExpression): String {
+        throw Exception("Unexpected current expression outside of flatten expression: $expression")
+    }
+
+    override fun visitExpressionType(expression: ExpressionTypeExpression): String {
+        throw Exception("ExpressionTypeExpression is unsupported")
+    }
+
+    override fun visitField(expression: FieldExpression): String {
+        return subfield(expression, "current")
```

Comment:
Like I mentioned in another comment, it'd be great if "current" was an instance property that way we can easily change the value to make generate code that is easier to read. (Giving it a default value of "current" is fine though)